packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +498,406 @@ class CommonFinders {
+  /// matching candidates.
+  ///
+  /// This is useful in cases where an action needs to be repeated while or
+  /// until a finder has results. The results from the search can be accessed
+  /// using the [found] property without re-executing the search.
+  bool tryFind() {
+    find();
+    return found.isNotEmpty;
+  }
+
+  /// Runs the given callback using cached results.
+  ///
+  /// While in this callback, this [FinderBase] will cache the next results from
+  /// [find] or [tryFind] and then no longer evaluate new results until the
+  /// callback completes. After the first call, all calls to [find], [tryFind]
+  /// or [found] will return the same results without evaluating.
+  void runCached(VoidCallback run) {
+    reset();
+    _cached = true;
+    try {
+      run();
+    } finally {
+      reset();
+      _cached = false;
+    }
+  }
+
+  /// Resets all state of this [FinderBase].
+  ///
+  /// Generally used between tests to reset the state of [found] if a finder is
+  /// used across multiple tests.
+  void reset() {
+    _found = null;
+  }
+
+  /// A string representation of this finder or its results.
+  ///
+  /// By default, this describes the results of the search in order to play
+  /// nicely with [expect] and its output when a failure occurs. If you wish
+  /// to get a string representation of the finder itself, pass [describeSelf]
+  /// as `true`.
+  @override
+  String toString({bool describeSelf = false}) {
+    if (describeSelf) {
+      return 'A finder that searches for ${describeCandidate(Plurality.many)}.';
+    } else {
+      if (!hasFound) {
+        find();
+      }
+      return found.toString();
+    }
+  }
+}
+
+/// The results of searching with a [FinderBase].
+class FinderResult<CandidateType> extends Iterable<CandidateType> {
+  /// Creates a new [FinderResult] that describes the [values] using the given
+  /// [describeCandidate] callback.
+  ///
+  /// {@macro flutter_test.finders.FinderBase.describeCandidate}
+  FinderResult(DescribeCandidateCallback describeCandidate, Iterable<CandidateType> values)
+    : _describeCandidate = describeCandidate, _values = values;
+
+  final DescribeCandidateCallback _describeCandidate;
+  final Iterable<CandidateType> _values;
+
+  @override
+  Iterator<CandidateType> get iterator => _values.iterator;
+
+  @override
+  String toString() {
+    final List<CandidateType> valuesList = _values.toList();
+    // This will put each value on its own line with a comma and indentation
+    final String valuesString = valuesList.fold(
+      '',
+      (String current, CandidateType candidate) => '$current\n  $candidate,',
+    );
+    return 'Found ${valuesList.length} ${_describeCandidate(valuesList.length.toPlurality())}: ['
+      '$valuesString'
+      '\n]';
+  }
+}
+
+/// Searches a widget tree and returns nodes that match a particular
+/// pattern.
+// TODO(pdblasi-google): Deprecate methods on this mixin.
```

Comment:
Deprecating the original Finder methods (now found in `_LegacyFinderMixin`) is primarily to improve the API, including the caching behavior.

These type names got updated to `*WidgetFinder` to make it explicit whether something was a Widget or Semantics finder. I'd probably push for that update regardless of any API changes to the classes themselves.

Reply:
Also, per the original question, I'm tracking the actual deprecations in #127136.

Reply:
It would be good to at least update the docs to point people to the correct replacement that they should be using instead to reduce some of the confusion around this. Having said that, I have some thoughts about the deprecations that I will leave as a separate top-level comment.

Reply:
Resolving and working on these concerns from notes from discussion over chat and the larger top-level comment.